HW1/P4/P4_bfs.py

```diff
@@ -0,0 +1,39 @@
+# Helper function - updates accumulator for newly explored nodes
+def nodes_update(x, explored):
+    explored.add(1)
+    return x
+
+# BFSS
+def bfs(sc, charGraph, startNode):
+    
+    # Explored depth of character graph
+    dist = 0
+
+    # Initial search nodes = starting node
+    searchNodes = sc.parallelize([startNode]).map(lambda x: (x, 0))
+
+    # Initial frontier = starting node
+    frontier = sc.parallelize([startNode]).map(lambda x: (x, 0))
+
+    # Accumulator to keep track of explored nodes
+    explored = sc.accumulator(1)
+    
+    # Keep going until there are no further nodes to expand
+    while not searchNodes.isEmpty():
+        
+        # Increment depth
+        dist += 1
+
+        # Look up all characters that are touched by all characters in the frontier
+        children = charGraph.join(searchNodes).flatMap(lambda x: x[1][0]).map(lambda x: (x, dist)).partitionBy(20)
```

Comment:
Did you avoid a shuffle here with the join? If you partition your RDD's before the join, you can avoid the huge costs.